src/bacnet/basic/object/csv.c

```diff
@@ -164,35 +183,90 @@ bool CharacterString_Value_Valid_Instance(uint32_t object_instance)
     unsigned index = 0; /* offset from instance lookup */
 
     index = CharacterString_Value_Instance_To_Index(object_instance);
-    if (index < MAX_CHARACTERSTRING_VALUES) {
+
+    if (object_instance == CSV_Descr[index].Instance) {
         return true;
     }
 
     return false;
 }
 
+/**
+ * Initialize the charactestring inputs. Returns false if there are errors.
+ *
+ * @param pInit_data pointer to initialisation values
+ *
+ * @return true/false
+ */
+bool CharacterString_Value_Set(BACNET_OBJECT_LIST_INIT_T *pInit_data)
+{
+  unsigned i;
+
+  if (!pInit_data) {
+    return false;
+  }
+
+  if ((int) pInit_data->length > MAX_CHARACTERSTRING_VALUES) {
+    PRINTF("pInit_data->length = %d >= %d", (int) pInit_data->length, MAX_CHARACTERSTRING_VALUES);
+    return false;
+  }
+
+  for (i = 0; i < pInit_data->length; i++) {
+    if (pInit_data->Object_Init_Values[i].Object_Instance < BACNET_MAX_INSTANCE) {
+        CSV_Descr[i].Instance = pInit_data->Object_Init_Values[i].Object_Instance;
+    } else {
+      PRINTF("Object instance %u is too big", pInit_data->Object_Init_Values[i].Object_Instance);
+      return false;
+    }
+
+    if(!strncmp(CSV_Descr[i].Name, pInit_data->Object_Init_Values[i].Object_Name, strlen(pInit_data->Object_Init_Values[i].Object_Name))) {
```

Comment:
why is statement below "Failed to set" if you're only comparing strings? I don't see a prior strncopy before this check in the same function.

Reply:
Thinking behind this is, if the instance isn't right, nothing that comes after will be right

Reply:
but the fields in your compare are the same in the copy. You're checking for Object Name, if it's not the same then print Fail, if it is the same then copy it over, except that you're copying over a value that's the same anyway. what is the point of the copy if the check already passes?

Reply:
Yeah, that makes complete sense. I've been just looking at it to long and was using the template for the other objects set functions for an example. I'll correct it